### ISSUE TITLE
Out of tree build broken

### ISSUE THREAD

**Issue by devinamatthews** [flame/blis#161]
Configuring and building from a directory other than the root source dir fails:

```
[dmatthews@twiddle sandybridge]$cd ~/build/blis/sandybridge
[dmatthews@twiddle sandybridge]$rm -rf *
[dmatthews@twiddle sandybridge]$~/src/blis/configure sandybridge
configure: reading configuration registry...done.
configure: checking whether we need to update the version file.
configure: checking version file '/home/dmatthews/src/blis/version'.
configure: starting configuration of BLIS 0.2.2.
configure: configuring with official version string.
configure: manual configuration requested; configuring with 'sandybridge'.
configure: checking configuration against contents of 'config_registry'.
configure: configuration 'sandybridge' is registered.
configure: 'sandybridge' is defined as having the following sub-configurations:
configure:    sandybridge
configure: which collectively require the following kernels:
configure:    sandybridge
configure: checking sub-configurations:
configure:   'sandybridge' is registered...and exists.
configure: checking sub-configurations' requisite kernels:
configure:   'sandybridge' kernels...exist.
configure: no install prefix given; defaulting to '/home/dmatthews/blis'.
configure: debug symbols disabled.
configure: disabling verbose make output, enable with 'make V=1'.
configure: building BLIS as a static library.
configure: threading is disabled.
configure: the BLAS compatibility layer is enabled.
configure: the CBLAS compatibility layer is disabled.
configure: the internal integer size is automatically determined.
configure: the BLAS/CBLAS interface integer size is 32-bit.
configure: creating ./config.mk from /home/dmatthews/src/blis/build/config.mk.in
configure: creating ./bli_config.h from /home/dmatthews/src/blis/build/bli_config.h.in
configure: creating ./obj/sandybridge
configure: creating ./obj/sandybridge/config
configure: creating ./obj/sandybridge/config/sandybridge
configure: creating ./obj/sandybridge/kernels
configure: creating ./obj/sandybridge/kernels/sandybridge
configure: creating ./obj/sandybridge/ref_kernels
configure: creating ./obj/sandybridge/ref_kernels/sandybridge
configure: creating ./obj/sandybridge/frame
configure: creating ./obj/sandybridge/testsuite
configure: creating ./lib/sandybridge
configure: creating ./include/sandybridge
configure: mirroring /home/dmatthews/src/blis/config/sandybridge to ./obj/sandybridge/config/sandybridge
configure: mirroring /home/dmatthews/src/blis/kernels/sandybridge to ./obj/sandybridge/kernels/sandybridge
configure: mirroring /home/dmatthews/src/blis/ref_kernels to ./obj/sandybridge/ref_kernels/sandybridge
configure: mirroring /home/dmatthews/src/blis/frame to ./obj/sandybridge/frame
configure: creating makefile fragment in /home/dmatthews/src/blis/config/sandybridge
configure: creating makefile fragment in /home/dmatthews/src/blis/kernels/sandybridge
configure: creating makefile fragment in /home/dmatthews/src/blis/kernels/sandybridge/3
configure: creating makefile fragment in /home/dmatthews/src/blis/ref_kernels
configure: creating makefile fragment in /home/dmatthews/src/blis/ref_kernels/1
configure: creating makefile fragment in /home/dmatthews/src/blis/ref_kernels/1f
configure: creating makefile fragment in /home/dmatthews/src/blis/ref_kernels/1m
configure: creating makefile fragment in /home/dmatthews/src/blis/ref_kernels/3
configure: creating makefile fragment in /home/dmatthews/src/blis/ref_kernels/ind
configure: creating makefile fragment in /home/dmatthews/src/blis/frame
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/absqsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/addsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/copysc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/divsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/getsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/mulsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/normfsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/setsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/sqrtsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/subsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/unzipsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/0/zipsc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/addv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/axpyv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/copyv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/dotv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/dotxv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/invertv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/kernels
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/packv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/scal2v
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/scalv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/setv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/subv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/swapv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1/unpackv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d/addd
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d/axpyd
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d/copyd
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d/invertd
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d/scal2d
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d/scald
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d/setd
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d/setid
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1d/subd
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1f
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1f/axpy2v
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1f/axpyf
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1f/dotaxpyv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1f/dotxaxpyf
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1f/dotxf
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1f/kernels
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/addm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/axpym
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/copym
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/packm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/packm/ukernels
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/scal2m
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/scalm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/setm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/subm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/unpackm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/1m/unpackm/ukernels
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/gemv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/ger
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/hemv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/her
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/her2
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/symv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/syr
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/syr2
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/trmv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/2/trsv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/gemm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/gemm/ind
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/gemm/ukernels
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/hemm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/her2k
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/herk
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/symm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/syr2k
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/syrk
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/trmm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/trmm3
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/trsm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/trsm/ukernels
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/3/ukernels
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/base
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/base/check
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/base/noopt
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/cntl
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/compat
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/compat/cblas
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/compat/cblas/f77_sub
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/compat/cblas/src
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/compat/check
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/compat/f2c
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/compat/f2c/util
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0/1e
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0/1m
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0/1r
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0/io
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0/ri
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0/ri3
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0/rih
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0/ro
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/include/level0/rpi
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/cntl
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/cntx
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/include
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/misc
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/oapi
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/query
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/tapi
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/ukernels
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/ukernels/gemm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/ind/ukernels/trsm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/thread
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/amaxv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/asumv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/mkherm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/mksymm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/mktrim
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/norm1m
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/norm1v
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/normfm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/normfv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/normim
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/normiv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/printm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/printv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/randm
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/randv
configure: creating makefile fragment in /home/dmatthews/src/blis/frame/util/sumsqv
configure: creating symbolic link to Makefile.
configure: creating symbolic link to common.mk.
configure: configured to build outside of source distribution.
[dmatthews@twiddle sandybridge]$make
Generating monolithic blis.h/bin/sh: build/flatten-headers.sh: No such file or directory
Makefile:370: recipe for target '/home/dmatthews/src/blis/include/sandybridge/blis.h' failed
make: *** [/home/dmatthews/src/blis/include/sandybridge/blis.h] Error 127
```

**Comment by fgvanzee** [comment]
Thanks Devin. This should be fixed now in 87978f6.

**Comment by devinamatthews** [comment]
Still broken:

```
[dmatthews@Devin's iMac release]$make
-n Generating monolithic blis.h
........................................................................................................................................................................................................................................................................................................................................................................................................................................................mv: rename /Users/dmatthews/src/blis/frame/include/blis.h.interm to /Users/dmatthews/src/blis/include/intel64/blis.h: No such file or directory
.
Generated /Users/dmatthews/src/blis/include/intel64/blis.h
Compiling obj/intel64/config/haswell/bli_cntx_init_haswell.o ('haswell' CFLAGS for config code)
Compiling obj/intel64/config/sandybridge/bli_cntx_init_sandybridge.o ('sandybridge' CFLAGS for config code)
Compiling obj/intel64/config/penryn/bli_cntx_init_penryn.o ('penryn' CFLAGS for config code)
Compiling obj/intel64/config/generic/bli_cntx_init_generic.o ('generic' CFLAGS for config code)
In file included from /Users/dmatthews/src/blis/config/generic/bli_cntx_init_generic.c:35:0:
/Users/dmatthews/src/blis/frame/include/blis.h:62:10: fatal error: bli_config.h: No such file or directory
 #include "bli_config.h"
          ^~~~~~~~~~~~~~
In file included from /Users/dmatthews/src/blis/config/sandybridge/bli_cntx_init_sandybridge.c:35:0:
/Users/dmatthews/src/blis/frame/include/blis.h:62:10: fatal error: bli_config.h: No such file or directory
 #include "bli_config.h"
          ^~~~~~~~~~~~~~
compilation terminated.
compilation terminated.
In file included from /Users/dmatthews/src/blis/config/haswell/bli_cntx_init_haswell.c:35:0:
/Users/dmatthews/src/blis/frame/include/blis.h:62:10: fatal error: bli_config.h: No such file or directory
 #include "bli_config.h"
          ^~~~~~~~~~~~~~
compilation terminated.
In file included from /Users/dmatthews/src/blis/config/penryn/bli_cntx_init_penryn.c:35:0:
/Users/dmatthews/src/blis/frame/include/blis.h:62:10: fatal error: bli_config.h: No such file or directory
 #include "bli_config.h"
          ^~~~~~~~~~~~~~
compilation terminated.
make: *** [obj/intel64/config/haswell/bli_cntx_init_haswell.o] Error 1
make: *** Waiting for unfinished jobs....
make: *** [obj/intel64/config/penryn/bli_cntx_init_penryn.o] Error 1
make: *** [obj/intel64/config/generic/bli_cntx_init_generic.o] Error 1
make: *** [obj/intel64/config/sandybridge/bli_cntx_init_sandybridge.o] Error 1
```

**Comment by fgvanzee** [comment]
@devinamatthews I can't immediately reproduce your error on Linux (as I could the first time). I'll try on OS X.

**Comment by fgvanzee** [comment]
@devinamatthews I couldn't reproduce on OS X either. What commit are you using?

**Comment by devinamatthews** [comment]
70a64432ee5a7adbee10fb7ff6d7b608c1940a7a

**Comment by ShadenSmith** [comment]
I just reproduced this using the latest commit on master (generic arch compiling on Arch Linux).

**Comment by fgvanzee** [comment]
I just noticed you are targeting `intel64`. Let me try that instead of `haswell`.

**Comment by devinamatthews** [comment]
@fgvanzee are you building out-of-tree *and* using a source dir that has not been also built in-tree?

**Comment by fgvanzee** [comment]
The key variable was your configure target. (You left out the configure line that you were executing, so it didn't occur to me at first that the way you were configuring could be affecting whether the error manifested.)

I'll dig into it.

**Comment by ShadenSmith** [comment]
On Linux and a fresh checkout of 70a64432, out-of-tree builds are failing for generic, intel64, and haswell configurations. I assume that it will fail for others.

**Comment by fgvanzee** [comment]
@ShadenSmith For some reason, I can't get it to fail for `haswell`.

Sigh. One thing at a time, I guess. I'll need to remind myself how out-of-tree builds are supported in the build system. (I never use them.)

**Comment by devinamatthews** [comment]
@fgvanzee I'm guessing you had already built `haswell` in-tree...

**Comment by ShadenSmith** [comment]
@fgvanzee Is it possible you're compiling with a source tree that has been built previously as @devinamatthews suggested? I'm working from a completely fresh checkout (never built in-tree before). Since it's failing for all of the configuration targets I try, I do not think it's a target-specific issue.

**Comment by fgvanzee** [comment]
Yes, I am building with a previously-built source tree. I misread Devin's original comment as having to do with a previously-built out-of-tree directory. Thanks for pointing this out.

**Comment by ShadenSmith** [comment]
Confirmed...I was able to build `haswell` out-of-tree only after first building it in-tree.

**Comment by fgvanzee** [comment]
BTW, can either of you remind me why out-of-tree builds are useful?

**Comment by devinamatthews** [comment]
Anytime you want to build more than one configuration or build with different flags (debug etc.), configure options, etc.

**Comment by fgvanzee** [comment]
Building more than one configuration is supported in-tree. But, only one configuration is "active" at any given time. So is it the switching between them without re-running `configure` that you're after?

**Comment by devinamatthews** [comment]
No.

**Comment by devinamatthews** [comment]
We should add an out-of-tree CI build while we're at it. (And an install test but that is another issue)

**Comment by fgvanzee** [comment]
I think 8d8ff74 completes the fix.

**Comment by devinamatthews** [comment]
Yes it seems to work now. The -n thing is still there on OSX but that is just cosmetic.
